staging/src/k8s.io/apiserver/pkg/endpoints/openapi/interfaces.go

```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+// Getter is an interface for fetching openapi specs and parsing them into an Resources struct
+type Getter interface {
+	// Get returns the entire OpenAPI schema
+	Get() (Resources, error)
+}
+
+// Resources interface describe a resources provider, that can give you
+// resource based on group-version-kind.
+type Resources interface {
+	// LookupResource looks up the OpenAPI schema for a particular gvk
+	LookupResource(gvk schema.GroupVersionKind) proto.Schema
+}
```

Comment:
Seems like we need one more thing, which binds a Resources to a particular resource.

```go
type ResourceSource interface {
  Resource() (proto.Schema, error)
}

type resourceGetter struct {
  getter Getter
  gvk schema.GroupVersionKind
}

func (r resourceGetter) Resource() (proto.Schema, error) {
  return r.getter.Get().LookupResource(r.gvk)
}
```

...some error handling left as an exercise to the reader ;)

Reply:
Yes I think we can need something like that too, to reduce the scope... But I wasn't sure if it should bind to a GVK or to just a Group, because I thought multiple types in a single api group can reference each other

Reply:
Also I like source better than getter as well, I just copied the Getter interface from kubectl as well.

Reply:
I had thought about doing an implementation of Resource which is exactly the same but won't allow for lookups outside a certain group, would that be enough to keep it in scope? It would still point tothe same getter and have the same Lookup(gvk) signature, but it would return nil unless the group matched some member variable we could set when creating that "restricted" resource

Reply:
something like 

```
// SchemaCacher fetches the openapi schema once and then caches it in memory
type restrictedSchemaSource struct {
	source       SchemaSource
	allowedGroup string
}

var _ SchemaSource = &restrictedSchemaSource{}

// NewSchemaSourceForGroup wraps SchemaSource so Resources objects returned by it
// will only allow lookups within a certain group.
func NewSchemaSourceForGroup(source SchemaSource, allowedGroup string) *restrictedSchemaSource {
	return &restrictedSchemaSource{
		source:       source
		allowedGroup: allowedGroup
	}
}

func (r *restrictedSchemaSource{}) Get() (Resources, error) {
	resources, err := r.source.Get()
	if err != nil {
		return nil, err
	}
	return &restrictedResources{
		resources: resources,
		group:     r.allowedGroup,
	}
}

type restrictedResources struct {
	resources    Resources
	allowedGroup string
}

var _ Resources = &restrictedResources{}

func (r *restrictedResources) LookupResource(gvk schema.GroupVersionKind) proto.Schema {
	if gvk.Group != r.allowedGroup {
		return nil
	}
	return r.resources.LookupResource(gvk)
}
```

Reply:
Hmm actually don't some resources reference resources outside their group?
apps/v1 referencing Pods, etc

Reply:
(we may have to walk the tree...)